parser/src/format/mod.rs

```diff
@@ -253,6 +255,102 @@ fn compression_from_filename(filename: &str) -> Option<Compression> {
+    inner: I,
+    /// The amount of errors we can absorb before halting parsing.
+    threshold: ErrorThreshold,
+    /// The number of records we've seen.
+    total_records: usize,
+    /// The number of errors we've seen lately.
+    errors_in_latest_window: usize,
+    /// The error messages from all the rows that have failed to parse.
+    absorbed: Vec<ParseError>,
+}
+
+impl<I: Iterator> ParseErrorBuffer<I> {
+    /// Consumes the next item out of the inner iterator.
+    fn advance(&mut self) -> Option<I::Item> {
+        // How many of the recent records to consider when trying to decide if
+        // we've entered a new region of bad data of the file.
+        const ERROR_BUFFER_WINDOW_SIZE: usize = 1000;
+
+        self.total_records += 1;
+
+        if self.total_records > ERROR_BUFFER_WINDOW_SIZE {
+            self.errors_in_latest_window = usize::saturating_sub(self.errors_in_latest_window, 1);
+        }
+
+        self.inner.next()
+    }
+
+    /// Saves the ParseError so it can be returned if we exceed our error threshold.
+    fn absorb(&mut self, e: ParseError) {
+        self.absorbed.push(e);
+        self.errors_in_latest_window += 1;
+    }
+
+    /// Returns true if we've exceeded the error threshold for the entire file.
+    fn is_full(&self) -> bool {
+        self.exceeded_limit(self.absorbed.len())
+    }
+
+    /// Returns true if we've exceeded the error threshold for the most recent records.
+    fn in_bad_region(&self) -> bool {
+        self.exceeded_limit(self.errors_in_latest_window)
+    }
+
+    fn exceeded_limit(&self, error_count: usize) -> bool {
+        // We haven't encountered any errors, so we certainly haven't exceeded our limit.
+        if self.total_records == 0 || error_count == 0 {
+            return false;
+        }
+
+        let error_rate = (error_count as f64 / self.total_records as f64 * 100.0) as u8;
+        self.threshold.exceeded(error_rate)
+    }
+}
+
+impl<I> ParseErrorBuffer<I> {
+    fn new(inner: I, threshold: ErrorThreshold) -> Self {
+        Self {
+            inner,
+            threshold,
+            total_records: 0,
+            errors_in_latest_window: 0,
+            absorbed: Vec::default(),
+        }
+    }
+}
+
+impl<I: Iterator<Item = ParseResult>> Iterator for ParseErrorBuffer<I> {
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.advance()? {
+                Ok(value) => return Some(Ok(value)),
+                Err(e) => {
+                    self.absorb(e);
+
+                    if self.is_full() || self.in_bad_region() {
+                        let limit = std::mem::take(&mut self.threshold);
+                        let errors = std::mem::take(&mut self.absorbed);
+                        return Some(Err(ParseError::ErrorLimitExceeded(limit, errors)));
+                    } else {
+                        tracing::error!(record_count = self.total_records, "failed to parse row");
```

Comment:
I'm thinking this log should be at the `warn` level, and should include some more information in the fields. Eventually, the connector can forward those logs to Flow, which can publish them to the logs collection. The ideal end state being that there's a warning log in their logs collection that has the filename, line number, and the actual content of any line that's triggering the error. I'm not suggesting that you need to do all that in this PR, but if you at least get the errors logged with sufficient information, then we can work on the other pieces.

Reply:
I waffled back and forth about `warn` vs `error`. It *is* an error, but we're skipping it. :upside_down_face: 

I wasn't sure how much info we'd want to output in the event we're skipping the row. I can log the entire error which includes specific span information. I can also add a bit more tracing context elsewhere so that this includes more information (the file name).

Reply:
I looked into this a bit more. It looks like we are already including the file name in the span information. 

Here's a sample of the log output for a malformed row:

```json
{
  "timestamp": "Sep 30 15:27:26.815",
  "level": "WARN",
  "fields": {
    "message": "failed to parse row",
    "item": "Parse(InvalidContent(Error(UnequalLengths { pos: Some(Position { byte: 664, line: 62, record: 61 }), expected_len: 2, len: 3 })))"
  },
  "target": "parser::format::csv",
  "span": {
    "parse_args": "ParseArgs { config_file: Some(\"/tmp/jsonl-parser-test.JcgEhXN5qFwT/config.json\"), file: \"-\" }",
    "name": "do_parse"
  }
}
```

If we pass the file as stdin (as with tests) you end up getting `"file": "-"`, but it should already be there.

Anyway, I think we can probably do a separate round of enriching this log data so that it's easier for an end user to consume, but we probably want to talk more about that.